packages/babel-helpers/src/helpers/applyDecs2311.ts

```diff
@@ -0,0 +1,596 @@
+          );
+        }
+      } else {
+        Object.defineProperty(Class, name, desc);
+      }
+    }
+    return newValue;
+  }
+
+  /* @no-mangle */
+  function applyMemberDecs(
+    Class: any,
+    decInfos: DecoratorInfo[],
+    instanceBrand: Function,
+    metadata: any,
+  ) {
+    var ret: Function[] = [];
+    var protoInitializers: Function[];
+    var staticInitializers: Function[];
+    var staticBrand = function (_: any) {
+      return checkInRHS(_) === Class;
+    };
+
+    var existingNonFields = new Map();
+
+    function pushInitializers(initializers: Function[]) {
+      if (initializers) {
+        ret.push(runInitializers.bind(null, initializers));
+      }
+    }
+
+    for (var i = 0; i < decInfos.length; i++) {
+      var decInfo = decInfos[i];
+
+      // skip computed property names
+      if (!Array.isArray(decInfo)) continue;
+
+      var kind = decInfo[1];
+      var name = decInfo[2];
+      var isPrivate = decInfo.length > 3;
+
+      var decoratorsHaveThis = kind & PROP_KIND.DECORATORS_HAVE_THIS;
+      var isStatic = !!(kind & PROP_KIND.STATIC);
+
+      kind &= 7; /* 0b111 */
+
+      var isField = kind === PROP_KIND.FIELD;
+      var isAccessor = kind === PROP_KIND.ACCESSOR;
+
+      var key = name + "/" + isStatic;
+
+      if (!isField && !isPrivate) {
+        var existingKind = existingNonFields.get(key);
+
+        if (
+          existingKind === true ||
+          (existingKind === PROP_KIND.GETTER && kind !== PROP_KIND.SETTER) ||
+          (existingKind === PROP_KIND.SETTER && kind !== PROP_KIND.GETTER)
+        ) {
+          throw new Error(
+            "Attempted to decorate a public method/accessor that has the same name as a previously decorated public method/accessor. This is not currently supported by the decorators plugin. Property name was: " +
+              name,
+          );
+        }
+        existingNonFields.set(key, kind > PROP_KIND.METHOD ? kind : true);
+      }
+
+      applyDec(
+        isStatic ? Class : Class.prototype,
+        decInfo,
+        decoratorsHaveThis,
+        isPrivate ? "#" + name : (toPropertyKey(name) as string),
+        kind,
+        metadata,
+        isField || isAccessor
+          ? /* fieldInitializers */ []
+          : isStatic
+            ? (staticInitializers = staticInitializers || [])
+            : (protoInitializers = protoInitializers || []),
+        ret,
+        isStatic,
+        isPrivate,
+        isField,
+        isAccessor,
+        isStatic && isPrivate ? staticBrand : instanceBrand,
+      );
+    }
+
+    pushInitializers(protoInitializers);
+    pushInitializers(staticInitializers);
+    return ret;
+  }
+
+  function defineMetadata(Class: any, metadata: any) {
+    return Object.defineProperty(
+      Class,
+      Symbol.metadata || Symbol.for("Symbol.metadata"),
+      { configurable: true, enumerable: true, value: metadata },
+    );
+  }
+
+  if (arguments.length >= 6) {
```

Comment:
I remember you removed this check.

Reply:
When `applyDecs2305` was wrapped into a factory, the check in `applyDecs2305.ts` was replaced because the wrapper always pass 6 parameters (see https://github.com/babel/babel/commit/a19712fba7890163c89c26df24f2a54fc8b7938a). Now that `applyDecs2305` does not have to be changed, this commit is removed as a result.

The check in `applyDecs2311` is copied from `applyDecs2305.ts`. We could update it to `parentClass !== void 0` as a class can not extend `undefined` and it will save us one byte or two.